### DISCUSSION TITLE
Error when use Promise .then(): Cannot modify '...', the object is protected and can only be modified by using an action.

### DISCUSSION THREAD

**Discussion by KyryloSWAN** [mobxjs/mobx-state-tree#2005] [Q&A] [answered]
I have array inside model:
```
import { types, Instance, flow, getParent } from "mobx-state-tree";
...
const Destination = types.model({
  addressLine1: types.maybe(types.string),
  id: types.identifier,
});
export const DestinationStore = types
  .model("Destinations", {
    destinations: types.optional(types.array(Destination), []),
  })
  .views((self) => ({
    get token(): string {
      return getParent<typeof RootStore>(self).token;
    },
  }))
  .actions((self) => {
    return {
      fetchDestinations: flow(function* getDestinations() {
        yield apiFetchDestinations(self.token).then((value) => {
          self.destinations.length = 0;
          value.data.destinations.forEach((item) => {
            self.destinations.push({
              addressLine1: item.addressLine1 ?? undefined,
              id: item.id,
            });
          });
        });
      }),
    };
  });
```
I'm always getting error `[mobx-state-tree] Cannot modify 'Destinations@/destination', the object is protected and can only be modified by using action fetchDestinations.`

I tried to rewrite the code from forEach to map:
```
const undefDestinations = value.data.destinations.map((item) =>
   ({
     addressLine1: item.addressLine1 ?? undefined,
     id: item.id,
   })
 );
self.destinations = cast([...undefDestinations]);
```
and got the same error.

Then I tried to add the action addAPIItem to Destination
```
.actions((self) => {
  return {
    addAPIItem(item: IAPIDestinationItem) {
      self.addressLine1 = item.addressLine1 ?? undefined;
      self.id = item.id;
    },
  };
});
```
but I couldn't find a way to call it from the action fetchDestinations.

If I don't use Promise .then no errors occur (but the code is bad, don't use it):
```
      fetchDestinations: flow(function* getDestinations() {
        const response = yield apiFetchDestinations(self.token);
        if (response.data?.destinations?.length) {
          response.data.destinations.forEach((item: any) => {
            item.addressLine1 ||= undefined;
          });
          self.destinations = response.data.destinations;
        }
      }),

```

I'm stuck with Promise  .then usage.
Please help!

**Comment by coolsoftwaretyler** [comment, marked as answer]
So I think when you chain off a promise with `.then()`, you turn the `.then()` clause into effectively an async function call.

That pulls the code execution _outside of the async generator you're using_, and [MST doesn't allow changing state outside actions](https://mobx-state-tree.js.org/concepts/async-actions#using-separate-actions)

You said your last code block is bad, but I think that's the correct thing to do here. Assuming `apiFetchDestinations` is some async function that returns a response (which would have been `value` in your `.then()` method), it should work. 

Can you say more about what's bad about your last block which yields the response value and uses it afterward? Or perhaps send a CodeSandbox to illustrate additional problems you're having?

**Reply by KyryloSWAN** [reply]
If we are using 
`const response = yield apiFetchDestinations(self.token);`
the type of `response` is uncertain and IDE shows it as `any` and we can't use the power of types checking.
But if we are using Promise `.then((item) => {...})`, the type of item is correct and contains type of successful response from the back-end server.

**Reply by coolsoftwaretyler** [reply]
Ah I see. I think in that case, you might just need to do two things:

1. You can use [TypeScript casting](https://www.w3schools.com/typescript/typescript_casting.php) to assert that the response is of a particular type.
2. You should be handling unsuccessful responses from your server with [try...catch blocks](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/try...catch), or checking for status codes, depending on how your backend chooses to handle things.

If you do that, you can satisfy TypeScript's expectations with regards to the data from the network, and you should be reasonably confident in the data at runtime as well - all while staying inside the generator function, which will allow you to modify state directly. 

Does that help? If you still run into problems, I think you'll want to send a CodeSandbox for clarity.

**Reply by coolsoftwaretyler** [reply]
For others who are curious about the TypeScript issue here, I found this: https://github.com/microsoft/TypeScript/issues/32523

It looks like TS has problems with type inference in generators. Lots of recommendations there about how to handle it, and I think typecasting here is basically appropriate.

The TS issue doesn't crop up in the `then()` operation because it's no longer a yielded value, and can infer the return type from the function appropriately.

Interesting stuff!

**Reply by OnkelTem** [reply]
Is there still a way to do it without a generator? Does it mean that one doesn't simply use regular promises inside actions?

Here is a sandbox where I get that exact error:
https://codesandbox.io/p/sandbox/discussion-2240-demo-forked-v2d3g8?workspaceId=ws_NT8YX4jqcz3FrmwfaKcnT5

```ts
const RootStore = t
  .model("RootStoreV2", {
    // Say we want to save a Map instead
    pets: t.map(Pet$),
  })
  .actions((self) => ({
    fetchPets: async () => {
      const pets = await getPets();
      pets.map((pet) => self.pets.set(pet.petId, pet)); // ERROR
    },
  }));

```

```
Uncaught Error: [mobx-state-tree] Cannot modify 'Map<string, Pet>@/pets', the object is protected
and can only be modified by using an action.
```

:-/

**Reply by OnkelTem** [reply]
Here is another attempt, even with `flow()` it somehow doesn't work:

```ts
// An auxiliary wrapper
const getPetsWrappedResp$ = t.model({
  pets: t.array(Pet$),
});

const RootStore = t
  .model("RootStore", {
    // Say we want to save a Map instead
    pets: t.map(Pet$),
  })
  .actions((self) => ({
    fetchPets: flow(function* () {
      // Getting whatever getPets() returns, untyped as we actually don't know the type (we're way picky to trust the API!)
      const untypedPets = yield getPets()
      // Wrap the result to match the shape of the getPetsWrappedResp$ model
      const wrappedRes = { pets: untypedPets }
      // validating API response and get the model instantiated
      const pets = getPetsWrappedResp$.create(wrappedRes);

      // Now let's build our own map so that we could access pets by their IDs
      pets.pets.map((pet) => {
        console.log({ pet });
        // Oops, somehow this doesn't work
        self.pets.set(pet.petId, pet)  // Error
      });
    }),
  }));
```

```
Uncaught Error: [mobx-state-tree] Cannot add an object to a state tree if it is already part of the same or another state tree. Tried to assign an object to '/pets/1', but it lives already at '/pets/0'
    at fail$1 (/node_modules/mobx-state-tree/dist/mobx-state-tree.module.js:3493:10)
```

[Sandbox](https://codesandbox.io/p/sandbox/discussion-2240-demo-forked-qlzxsn?workspaceId=ws_NT8YX4jqcz3FrmwfaKcnT5)

**Reply by OnkelTem** [reply]
Seems like I could get it working in [another sandbox](https://codesandbox.io/p/sandbox/discussion-2240-demo-forked-n63vmp?workspaceId=ws_NT8YX4jqcz3FrmwfaKcnT5)

Two changes I made:

1. `pets: t.map(Pet$),` -> `pets: t.map(t.reference(Pet$)),`
2. `petId: t.number,` -> `petId: t.identifierNumber,`

Does it look proper now? An odd thing is that now snapshot contains only ids:
```
Pets after map action:
{1: 1, 2: 2}1: 12: 2[[Prototype]]: Object
```

**Reply by coolsoftwaretyler** [reply]
> Is there still a way to do it without a generator? Does it mean that one doesn't simply use regular promises inside actions?

Yes, you can do this without generators, but when you manipulate the tree, you _need_ to use actions in every case. And if you chain off another promise in `then`, you will _need_ to use actions still. [Here is a blog post about it](https://coolsoftware.dev/blog/write-async-await-with-mobx-state-tree/).

> Here is another attempt, even with flow() it somehow doesn't work:

In this attempt, you took your API response and added it to a tree in `getPetsWrappedResp$` . [Here is the documentation about our tree semantics](https://mobx-state-tree.js.org/concepts/trees#tree-semantics-in-detail). References help here because they allow you refer to models in a different part of the tree.

But if `getPetsWrappedResp$` is just responsible for runtime type checking, there's no reason you couldn't just allow your `t.map` to do that:

```ts
import { t } from "mobx-state-tree";

(async function () {
  const Pet = t.model("Pet", {
    name: t.string,
  });

  const RootStore = t
    .model("RootStore", {
      pets: t.map(Pet),
    })
    .actions((self) => ({
      correctSet() {
        const pet = { name: "dog" };
        self.pets.set(0, pet);
      },
      incorrectSet() {
        const notAPet = { wrongField: true };
        // @ts-expect-error - for demo
        self.pets.set(1, notAPet);
      },
    }));

  const store = RootStore.create();

  store.correctSet();
  console.log(store.pets);
  store.incorrectSet(); // Uncaught Error: [mobx-state-tree] Error while converting `{"wrongField":true}` to `Pet`:
})();
```

[See in codesandbox](https://codesandbox.io/p/sandbox/68y8dr?file=%2Fsrc%2Findex.ts%3A30%2C1).

This is another benefit of MST - you get author-time type checking with our TypeScript types, and runtime type checking (although we turn this behavior off in production for performance, so you can [also use the typecheck utility if you want it explicitly at runtime](https://mobx-state-tree.js.org/API/#typecheck))

Your example with references looks fine, but I think you can simplify it by removing that other array. If you want both the array and the map for reasons other than type checking, then you'll want references in one of them (assuming you want to share the exact same instance. You could also duplicate the data without issue)

**Comment by KyryloSWAN** [comment]
I've modified action code by your suggestions and it works!
```
      fetchDestinations: flow(function* getDestinations() {
        self.items.length = 0;
        self.isLoading = true;
        try {
          const response: IAPIDestinationsResponse = yield apiFetchDestinations(
            self.token
          );
          response?.data?.destinations?.forEach((item: IAPIDestinationItem) => {
            self.items.push({
              addressLine1: item.addressLine1 ?? undefined,
              id: item.id,
            });
          });
        } finally {
          self.isLoading = false;
        }
      }),
```
Types checking works and I didn't receive any error message :)
Thank you for your support @coolsoftwaretyler
